lib/Sema/TypeCheckPropertyWrapper.cpp

```diff
@@ -174,16 +174,31 @@ static ConstructorDecl *findInitialValueInit(ASTContext &ctx,
 static ConstructorDecl *findDefaultInit(ASTContext &ctx,
                                         NominalTypeDecl *nominal) {
   SmallVector<ConstructorDecl *, 2> defaultValueInitializers;
-  DeclName initName(ctx, DeclBaseName::createConstructor(),
-                    ArrayRef<Identifier>());
   SmallVector<ValueDecl *, 2> decls;
+  auto initName = DeclBaseName::createConstructor();
   nominal->lookupQualified(nominal, initName, NL_QualifiedDefault, decls);
   for (const auto &decl : decls) {
     auto init = dyn_cast<ConstructorDecl>(decl);
     if (!init || init->getDeclContext() != nominal)
       continue;
 
-    defaultValueInitializers.push_back(init);
+    // A constructor which does not have any parameters or where all the
+    // parameters have a default argument can be used to default initialize
+    // the property wrapper type.
+    assert(init->hasParameterList());
+    auto hasParams = init->getParameters()->size() > 0;
+    auto allParamsHaveDefaultArg = false;
+
+    if (hasParams) {
+      allParamsHaveDefaultArg = llvm::all_of(
+          init->getParameters()->getArray(),
+          [](const ParamDecl *decl) { return decl->isDefaultArgument(); });
+    }
+
+    // Skip synthesized default initializers
+    if (!hasParams || (allParamsHaveDefaultArg && !init->isImplicit())) {
```

Comment:
Why skip the synthesized default initializer? It's reasonable to consider calling it, so long as it's internal.

Reply:
We might have a init() and a synthesised init(wrappedValue:), which now triggers the “multiple default value initialiser” unless we ignore the synthesised init.

Reply:
Hmm. We could use `TypeChecker::compareDeclarations` to pick the best default-value initializer. That seems more principled than ignoring implicit inits.

Reply:
Sounds like a good idea! I have done that now and it seems to work as expected.

Reply:
What makes something `init1` versus `init3`? Source order? That would be brittle.

Seems to me that in the rare circumstance of many equally acceptable initializers, it would be reasonable to have the user explicitly clarify.

Reply:
The initializers are compared using `Typechecker::compareDeclarations`, which returns `Comparison::Unordered` if two decls are “equal” (they’re not better or worse than each other). See CSRanking.cpp to understand how it works.

I could bail out if we ever end up in that situation, instead of being greedy and using whatever “best” initializer we’ve found so far.